### PR TITLE
[SYCL] Enable discard_events mode for the Level Zero

### PR DESCRIPTION
**Pull request by againull**
* Teach the Level Zero plugin piEnqueue* functions to accept nullptr instead of a pointer to the output event. In this case event is created internally and is not visible externally.
* Introduce RefCountExternal for pi_event which allows to track external references to an event. It allows to do some optimizations if we know that event is not externally visible. These optimizations are going to be implemented in subsequent PRs.
* Don't create proxy event for batch if there are no externally visible events in the batch.

### CODE REVIEW

**Comment by againull** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1132

**Comment by againull** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1132

**Comment by againull** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1132

**Comment by againull** [comment]
[SYCL/Plugin/level_zero_events_caching.cpp](https://github.com/intel/llvm-test-suite/pull/1138/files#diff-d6364c1be6f81d8a01e5f5ce641014045a2e43fd5f4f0aeb05c251c3de560666) failure is unrelated to this PR, fix is uploaded for review here: https://github.com/intel/llvm-test-suite/pull/1138

HIP backend failures are unrelated because changes are isolated to L0 plugin.

File: sycl/include/sycl/detail/pi.h [hunk 1]
```diff
@@ -564,6 +564,8 @@ constexpr pi_queue_properties PI_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE = (1 << 0);
 constexpr pi_queue_properties PI_QUEUE_PROFILING_ENABLE = (1 << 1);
 constexpr pi_queue_properties PI_QUEUE_ON_DEVICE = (1 << 2);
 constexpr pi_queue_properties PI_QUEUE_ON_DEVICE_DEFAULT = (1 << 3);
+constexpr pi_queue_properties PI_EXT_ONEAPI_QUEUE_DISCARD_EVENTS_MODE_ENABLE =
```
Comment (smaslov-intel) [inline sycl/include/sycl/detail/pi.h:567]: this requires a change of the version, see the header of this file

Comment (againull) [reply sycl/include/sycl/detail/pi.h:569]: Thanks, changed the version.

**Comment by againull** [comment]
@smaslov-intel Could you please review this PR?

File: sycl/include/sycl/detail/pi.h [hunk 2]
```diff
@@ -564,6 +566,8 @@ constexpr pi_queue_properties PI_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE = (1 << 0);
...
```
Comment (smaslov-intel) [inline sycl/include/sycl/detail/pi.h:569]: I think the mode-enable doesn't convey any additional info, let's shorten this to:
```suggestion
constexpr pi_queue_properties PI_EXT_ONEAPI_QUEUE_DISCARD_EVENTS =
```

File: sycl/source/detail/queue_impl.hpp [hunk 3]
```diff
@@ -96,11 +96,8 @@ class queue_impl {
         MDiscardEvents(
             has_property<ext::oneapi::property::queue::discard_events>()),
         MIsProfilingEnabled(has_property<property::queue::enable_profiling>()),
-        MHasDiscardEventsSupport(
-            MDiscardEvents &&
-            (MHostQueue ? true
-                        : (MIsInorder && getPlugin().getBackend() !=
-                                             backend::ext_oneapi_level_zero))) {
+        MHasDiscardEventsSupport(MDiscardEvents &&
```
Comment (smaslov-intel) [inline sycl/source/detail/queue_impl.hpp:99]: Why is discard-event only supported fir in-order?

File: sycl/source/detail/queue_impl.hpp [hunk 4]
```diff
@@ -317,6 +311,13 @@ class queue_impl {
             ext::oneapi::cuda::property::queue::use_default_stream>()) {
       CreationFlags |= __SYCL_PI_CUDA_USE_DEFAULT_STREAM;
     }
+    if (getPlugin().getBackend() == backend::ext_oneapi_level_zero &&
+        MPropList
+            .has_property<ext::oneapi::property::queue::discard_events>()) {
+      // Pass this flag to the Level Zero plugin to be able to check it from
```
Comment (smaslov-intel) [inline sycl/source/detail/queue_impl.hpp:317]: Why is this L0 specific? Why should we not pass it to the other plugins? (I think we should)

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 5]
```diff
@@ -5322,8 +5345,11 @@ piEnqueueKernelLaunch(pi_queue Queue, pi_kernel Kernel, pi_uint32 WorkDim,
     return Res;
 
   ze_event_handle_t ZeEvent = nullptr;
+  pi_event InternalEvent;
+  bool IsInternal = OutEvent == nullptr;
+  pi_event *Event = OutEvent ? OutEvent : &InternalEvent;
   pi_result Res = createEventAndAssociateQueue(
-      Queue, Event, PI_COMMAND_TYPE_NDRANGE_KERNEL, CommandList);
+      Queue, Event, PI_COMMAND_TYPE_NDRANGE_KERNEL, CommandList, IsInternal);
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:5352]: what is this pattern (used in many places) aimed to do? (besides obvious setting of the new "Internal" property of the pi_event)

it looks dangerous that you are using an unallocated (dangling) pointer to "InternalEvent"

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:5352]: I see that "Event" is actually an output parameter so it is will be allocated inside the call, OK then

File: sycl/source/detail/queue_impl.hpp (hunk 3 above)
Comment (againull) [reply sycl/source/detail/queue_impl.hpp:99]: I think we can enable support of discard-event for out-of-order queues as well, I can't think of a reason against this idea.
But it is not an intention of this PR so I will prefer to do this separately because it may expose unexpected issues or cause some long discussion which will delay merging of this PR which is currently blocking another PR.

File: sycl/source/detail/queue_impl.hpp (hunk 4 above)
Comment (againull) [reply sycl/source/detail/queue_impl.hpp:315]: Fixed to pass this property to all backends.

File: sycl/include/sycl/detail/pi.h (hunk 2 above)
Comment (againull) [reply sycl/include/sycl/detail/pi.h]: Fixed.

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 6]
```diff
@@ -1374,6 +1377,10 @@ struct _pi_event : _pi_object {
   // being visible to the host at all.
   bool Completed = {false};
 
+  // Indicates that event is internal, i.e. it is visible inside the L0 plugin
+  // only.
+  bool Internal = {false};
```
[the remark this answers is not in the document]
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.hpp]: If we do optimization (2-event model) only for discarded events in in-order queue then bool variable seems enough. But if we want to extend optimization for events which are not discarded then we need this ref count of external references to know when event is not externally visible anymore. 
So I used the ext ref count approach as suggested.

File: sycl/source/detail/queue_impl.hpp (hunk 3 above)
Comment (smaslov-intel) [reply sycl/source/detail/queue_impl.hpp:99]: OK

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 7]
```diff
@@ -686,7 +690,7 @@ inline static pi_result createEventAndAssociateQueue(
   // Append this Event to the CommandList, if any
   if (CommandList != Queue->CommandListMap.end()) {
     CommandList->second.append(*Event);
-    PI_CALL(piEventRetain(*Event));
+    (*Event)->RefCount.increment();
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:693]: Just curious: why is this change needed?

**Review by smaslov-intel — APPROVED** [review]
LGTM

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 7 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp:693]: piEventRetain increments external ref count, but we need to increment only internal ref count here. Same logic as for pi_queue.
